pinot-core/src/main/java/org/apache/pinot/core/data/readers/PinotSegmentRecordReader.java

```diff
@@ -72,7 +73,10 @@ public PinotSegmentRecordReader(@Nonnull File indexDir, @Nullable Schema schema,
       SegmentMetadata segmentMetadata = _immutableSegment.getSegmentMetadata();
       _numDocs = segmentMetadata.getTotalRawDocs();
       if (schema == null) {
-        _schema = segmentMetadata.getSchema();
+        // In order not to expose virtual columns to client, schema shouldn't be fetched from segmentMetadata;
+        // otherwise the original metadata will be modified.
+        _schema = new SegmentMetadataImpl(indexDir).getSchema();
+        _schema.getAllFieldSpecs().removeIf(fieldSpec -> _schema.isVirtualColumn(fieldSpec.getName()));
```

Comment:
You don't need to remove virtual columns here because virtual columns will not be added to the segment metadata without explicitly adding them.